### DISCUSSION TITLE
Text overlapping on addRedactAnnot and apply_redactions

### DISCUSSION THREAD

**Discussion by MallRoy** [pymupdf/PyMuPDF#1171] [Upstream Bugs]
I am facing a text overlapping issue on applying redactions in a pdf.
I am using python 3.9.2 and pip 20.2.3 versions.
Image before applying redaction
![image](https://user-images.githubusercontent.com/54269166/119232446-0372f300-bb43-11eb-8f06-b5555b275d72.png)

Image After applying redaction
![image](https://user-images.githubusercontent.com/54269166/119232429-df171680-bb42-11eb-9a3c-1ccb72737684.png)

I am using the below code for redaction

```
# imports
import fitz
import re
import sys
import json
from matplotlib import colors


class Redactor:
    # constructor
    def __init__(self, path, desc, out, vertexes, color):

        self.path = path
        self.desc = desc.split(',')
        self.out = out
        self.vertexes = vertexes
        self.color = color

    def redaction(self):

        # opening the pdf
        doc = fitz.open(self.path)
        # iterating through pages
        for page in doc:

            if not page._isWrapped:
                page.wrapContents()

            text = page.getText('text')
            currentpage = page.number
            dl = page.getDisplayList()
            tp = dl.getTextPage()
            # print(self.vertexes[currentpage])
            for searchtext in self.desc or []:
                # print(searchtext)
                areas = tp.search(searchtext, quads=True)
                # print(areas)
                if (areas is not None):
                    [page.addRedactAnnot(area, fill=colors.to_rgb(
                        self.color)) for area in areas]
                    page.apply_redactions()
                    # print(areas)

            for vertices in self.vertexes[currentpage]["vertices"]:
                # print(vertices)
                clip = fitz.Quad(((vertices[0]["x"], vertices[0]["y"]), (vertices[1]["x"], vertices[1]["y"]), (
                    vertices[2]["x"], vertices[2]["y"]), (vertices[3]["x"], vertices[3]["y"])))
                page.addRedactAnnot(clip, fill=colors.to_rgb(self.color))
                page.apply_redactions()

        # saving it to a new pdf
        doc.save(self.out, deflate=True, clean=True, linear=True)
        print("Successfully redacted")


# driver code for testing
if __name__ == "__main__":

    # replace it with name of the pdf file
    path = sys.argv[1]
    desc = sys.argv[2]
    out = sys.argv[3]
    color = sys.argv[4]
    lines = sys.stdin.readlines()
    gcpvalues = json.loads(lines[0])
    vertexes = gcpvalues["vertex"]
    redactor = Redactor(path, desc, out, vertexes, color)
    redactor.redaction()

```

**Comment by JorjMcKie** [comment]
Cannot recreate problem.
This could be cause by all sorts of things. Need at least a minimal PDF page for which this happens.

**Comment by JorjMcKie** [comment]
In the meantime you could try and unconditionally replace 
```python
            if not page._isWrapped:
                page.wrapContents()
```
by ``page.clean_contents()``.

**Comment by MallRoy** [comment]
> In the meantime you could try and unconditionally replace
> 
> ```python
>             if not page._isWrapped:
>                 page.wrapContents()
> ```
> 
> by `page.clean_contents()`.

We have tried this already, it is not working.

**Comment by JorjMcKie** [comment]
Hm, ok.
Then give me an example file for reproducing this please.

**Comment by JorjMcKie** [comment]
Use my e-mail address if any confidentiality concerns.

**Comment by MallRoy** [comment]
> Use my e-mail address if any confidentiality concerns.

sure

**Comment by JorjMcKie** [comment]
Thanks for the file.
This is an upstream (MuPDF) error. If you agree, I will send them the file and a reproducer program.
Your file does not seem to contain confidential information, so that shouldn't be a problem.
Apart from the fact that these people are absolutely trustworthy.

**Comment by MallRoy** [comment]
> Thanks for the file.
> This is an upstream (MuPDF) error. If you agree, I will send them the file and a reproducer program.
> Your file does not seem to contain confidential information, so that shouldn't be a problem.
> Apart from the fact that these people are absolutely trustworthy.

Ok. I have removed the sensitive content. So I don't have any concern about that part.

**Comment by JorjMcKie** [comment]
Good - I will open an issue at their system then and forward the link to you.

**Comment by MallRoy** [comment]
> Good - I will open an issue at their system then and forward the link to you.

ok

**Comment by JorjMcKie** [comment]
Here is the link: https://bugs.ghostscript.com/show_bug.cgi?id=703894

**Comment by MallRoy** [comment]
> Here is the link: https://bugs.ghostscript.com/show_bug.cgi?id=703894

Thankyou @JorjMcKie

**Comment by JorjMcKie** [comment]
You can open an accunt on bugzilla and enter your e-mail address as to be notified when there are status changes of the issue.

**Comment by MallRoy** [comment]
@JorjMcKie Status of the bug is resolved and fixed https://bugs.ghostscript.com/show_bug.cgi?id=703894#c0

**Reply by JorjMcKie** [reply]
Thank you - yes I have been notified too.
It will take effect in PyMuPMDF with the version based on the new release of MuPDF.
